doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,466 @@ It does not intend to describe the **current behavior** of the system.
+versions, and two frontiers: *since* and *upper*: `ptvc = [timeline,
+collection-versions, since, upper]`.
+
+Informally, the `since` frontier tells us which versions have been lost to the
+past, and the `upper` frontier tells us which versions have yet to be learned.
+At times between `since` and `upper`, a pTVC should be identical to its
+corresponding TVC.
+
+![A diagram showing a pTVC as a small region of a larger
+TVC](assets/tvc-ptvc.jpeg)
+
+We think of `since` as the "read frontier": times not later than or equal to
+`since` cannot be correctly read. We think of `upper` as the "write frontier":
+times later than or equal to `upper` may still be written to the TVC.
+
+A *read* of a pTVC `ptvc` at time `t` is the collection version in `ptvc` which
+corresponds to `t`. We write this as `read(ptvc, t) => collection-version`.
+
+We say that a read from a pTVC is *correct* when it is exactly the same
+collection version as would be read from the pTVC's corresponding TVC:
+`read(ptvc, t) = read(tvc, t)`.
+
+All pTVCs ensure an important correctness invariant. Consider a pTVC `ptvc`
+corresponding to some TVC `tvc`, and with `since` and `upper` frontiers. For
+all times `t` such that `t` is later than or equal to `since`, and `t` is *not*
+later than or equal to `upper`, reads of `ptvc` at `t` are correct: `read(ptvc,
+t) = read(tvc, t)`. In the context of a pTVC, we say that such a time `t`
+is *between* `since` and `upper`.
+
+![A diagram showing past, readable, and future times in a
+pTVC.](assets/since-readable-upper.jpeg)
+
+In this diagram, times are shown as a two-dimensional lattice forming a partial
+order; time generally flows left to right, and the precise order is shown by
+arrows. The `since` frontier is formed by two times labeled "since"; the
+`upper` frontier is formed by two times labeled "upper". Times later than
+`upper` are drawn in orange: these times are in the unknown future of the pTVC.
+Times not later than or equal to `since` are drawn in gray: these times are in
+the forgotten past. Only times in blue--those between `since` and `upper`--can
+be correctly read.
+
+## Update-set representations of TVCs and pTVCs
+
+Because the set of times is in general very large, and because TVCs usually
+change incrementally, it is usually more efficient to represent a TVC as a set
+of *updates* which encode changes between collection versions. An *update* is a
+triple `[data, time, diff]`. The `data` specifies what element of the
+collection version is changing, the `time` gives the time the change took
+place, and the `diff` is the change in multiplicity for that element.
+
+Given a collection version `cv1` and an update `u = [data, time, diff]`, we can
+*apply* `u` to `cv1` to produce a new collection version `cv2`: `apply(cv1, u)
+=> cv2`. `cv2` is exactly `cv1`, but with `multiplicity(cv2, data) =
+multiplicity(cv1, data) + diff`.
+
+We can also apply a countable set of updates to a collection version by
```

Comment:
I'm not sure how countably infinite sets of updates can work, unless you restrict the subset of updates at each piece of data to be finite.

For example, consider the set of updates [t0, "SomeData", n] for all positive integers n. The multiplicity of "SomeData" in the result of applying these to any collection version will grow without bound, but our definitions don't allow multiplicities to be infinite. (Or, even if they do, let the diffs in that set be n * (-1) ^ n ).

Reply:
Yeah, infinite multiplicities are a reasonable thing--Reisig defines them in Z. I just said "countable" here because I didn't want to get into whatever the weird Mobius transform pseudointegral was--that math is beyond me. In practice I don't think this matters: we're dealing with computers with finite space and time.